File: src/web/src/types/voice.ts
/**
 * Voice-related TypeScript interfaces, types, and enums for the Personal AI Agent.
 * Defines types for speech-to-text transcription, text-to-speech synthesis,
 * voice state management, and voice control components.
 */

/**
 * Enum representing the current state of voice interaction
 */
export enum VoiceState {
  IDLE = 'idle',
  LISTENING = 'listening',
  PROCESSING = 'processing',
  SPEAKING = 'speaking',
  ERROR = 'error'
}

/**
 * Enum for speech-to-text model options
 * Based on OpenAI Whisper models - version 20231117
 */
export enum SpeechModel {
  WHISPER_TINY = 'whisper-tiny',
  WHISPER_BASE = 'whisper-base',
  WHISPER_SMALL = 'whisper-small',
  WHISPER_MEDIUM = 'whisper-medium',
  WHISPER_LARGE = 'whisper-large'
}

/**
 * Enum for audio format options
 */
export enum AudioFormat {
  MP3 = 'mp3',
  WAV = 'wav',
  OGG = 'ogg'
}

/**
 * Interface for speech-to-text transcription requests
 */
export interface TranscriptionRequest {
  /**
   * The speech-to-text model to use
   */
  model: SpeechModel;
  
  /**
   * ISO 639-1 language code (e.g., 'en', 'fr', 'es')
   * If not specified, the model will attempt to detect the language
   */
  language?: string;
  
  /**
   * Optional prompt to guide the transcription
   */
  prompt?: string;
  
  /**
   * Model temperature, controls randomness. Range: 0.0 to 1.0
   * Lower values for more focused and deterministic outputs
   */
  temperature?: number;
  
  /**
   * Whether to automatically detect the language of the audio
   */
  detect_language?: boolean;
}

/**
 * Interface for speech-to-text transcription responses
 */
export interface TranscriptionResponse {
  /**
   * The transcribed text
   */
  text: string;
  
  /**
   * Detected or specified language of the audio
   */
  language: string;
  
  /**
   * Duration of the processed audio in seconds
   */
  duration: number;
  
  /**
   * Confidence score of the transcription (0.0 to 1.0)
   */
  confidence: number;
  
  /**
   * Array of segments with timestamps
   */
  segments?: Array<{
    start: number; // Start time in seconds
    end: number;   // End time in seconds
    text: string;  // Text in this segment
  }>;
}

/**
 * Interface for text-to-speech synthesis requests
 */
export interface SynthesisRequest {
  /**
   * ID of the voice to use for synthesis
   */
  voice_id: string;
  
  /**
   * Speech speed modifier (default: 1.0)
   * Range typically from 0.5 (slower) to 2.0 (faster)
   */
  speed?: number;
  
  /**
   * Voice pitch modifier (default: 1.0)
   * Range typically from 0.5 (lower) to 2.0 (higher)
   */
  pitch?: number;
  
  /**
   * Output audio format
   */
  format?: AudioFormat;
  
  /**
   * TTS provider to use (e.g., 'elevenlabs', 'native', 'coqui')
   */
  provider?: string;
}

/**
 * Interface for text-to-speech synthesis responses
 */
export interface SynthesisResponse {
  /**
   * The synthesized audio data as a Blob
   */
  audio_data: Blob;
  
  /**
   * Duration of the audio in seconds
   */
  duration: number;
  
  /**
   * Format of the audio data
   */
  format: AudioFormat;
  
  /**
   * ID of the voice used for synthesis
   */
  voice_id: string;
}

/**
 * Interface for voice information
 */
export interface VoiceInfo {
  /**
   * Unique identifier for the voice
   */
  id: string;
  
  /**
   * Display name of the voice
   */
  name: string;
  
  /**
   * Provider of the voice (e.g., 'elevenlabs', 'native', 'coqui')
   */
  provider: string;
  
  /**
   * Language of the voice as ISO 639-1 code
   */
  language: string;
  
  /**
   * Gender of the voice ('male', 'female', 'neutral')
   */
  gender: string;
  
  /**
   * URL to a preview audio clip of the voice
   */
  preview_url?: string;
  
  /**
   * Description of the voice characteristics
   */
  description?: string;
}

/**
 * Interface for requesting available voices
 */
export interface VoiceListRequest {
  /**
   * Optional provider to filter voices by
   */
  provider?: string;
  
  /**
   * Optional language code to filter voices by
   */
  language?: string;
  
  /**
   * Whether to refresh the cached voice list
   */
  force_refresh?: boolean;
}

/**
 * Interface for response containing available voices
 */
export interface VoiceListResponse {
  /**
   * Array of available voices
   */
  voices: VoiceInfo[];
}

/**
 * Interface for microphone state information
 */
export interface MicrophoneState {
  /**
   * Whether the microphone is currently recording
   */
  isRecording: boolean;
  
  /**
   * Whether recording is paused
   */
  isPaused: boolean;
  
  /**
   * Current audio level (0.0 to 1.0)
   */
  audioLevel: number;
  
  /**
   * Duration of the current recording in seconds
   */
  duration: number;
  
  /**
   * Error object if there's an issue with the microphone
   */
  error: Error | null;
}

/**
 * Interface for audio player state information
 */
export interface AudioPlayerState {
  /**
   * Whether audio is currently playing
   */
  isPlaying: boolean;
  
  /**
   * Whether audio playback is paused
   */
  isPaused: boolean;
  
  /**
   * Total duration of the current audio in seconds
   */
  duration: number;
  
  /**
   * Current playback position in seconds
   */
  currentTime: number;
  
  /**
   * Current volume level (0.0 to 1.0)
   */
  volume: number;
  
  /**
   * Error object if there's an issue with the audio player
   */
  error: Error | null;
}

/**
 * Interface for voice control component props
 */
export interface VoiceControlProps {
  /**
   * Callback function when transcription is complete
   * @param text The transcribed text
   */
  onTranscription: (text: string) => void;
  
  /**
   * Whether the voice control is disabled
   */
  disabled?: boolean;
  
  /**
   * CSS class name for styling
   */
  className?: string;
  
  /**
   * Whether to show tooltips for the voice control
   */
  showTooltip?: boolean;
  
  /**
   * Size of the voice control ('small', 'medium', 'large')
   */
  size?: 'small' | 'medium' | 'large';
  
  /**
   * Options for speech-to-text transcription
   */
  transcriptionOptions?: Partial<TranscriptionRequest>;
}

/**
 * Interface for voice hook options
 */
export interface VoiceHookOptions {
  /**
   * Whether to automatically start listening on hook initialization
   */
  autoStart?: boolean;
  
  /**
   * Options for speech-to-text transcription
   */
  transcriptionOptions?: Partial<TranscriptionRequest>;
  
  /**
   * Options for text-to-speech synthesis
   */
  synthesisOptions?: Partial<SynthesisRequest>;
  
  /**
   * Callback for voice state changes
   */
  onStateChange?: (state: VoiceState) => void;
  
  /**
   * Callback for when transcription is complete
   */
  onTranscription?: (text: string) => void;
  
  /**
   * Callback for errors
   */
  onError?: (error: Error) => void;
}

/**
 * Interface for voice hook result
 */
export interface VoiceHookResult {
  /**
   * Current state of voice interaction
   */
  state: VoiceState;
  
  /**
   * Whether the system is currently listening for speech
   */
  isListening: boolean;
  
  /**
   * Whether the system is currently speaking
   */
  isSpeaking: boolean;
  
  /**
   * Whether the system is processing audio
   */
  isProcessing: boolean;
  
  /**
   * Error object if there's an issue
   */
  error: Error | null;
  
  /**
   * Current transcription text
   */
  transcript: string;
  
  /**
   * Current audio input level (0.0 to 1.0)
   */
  audioLevel: number;
  
  /**
   * Function to start listening for speech
   */
  startListening: () => Promise<void>;
  
  /**
   * Function to stop listening and process the speech
   */
  stopListening: () => Promise<void>;
  
  /**
   * Function to cancel listening without processing
   */
  cancelListening: () => void;
  
  /**
   * Function to convert text to speech and play it
   * @param text The text to speak
   * @param options Optional synthesis options
   */
  speak: (text: string, options?: Partial<SynthesisRequest>) => Promise<void>;
  
  /**
   * Function to stop speaking
   */
  stopSpeaking: () => void;
  
  /**
   * Whether voice functionality is supported in the current environment
   */
  isSupported: boolean;
  
  /**
   * Array of available voices
   */
  availableVoices: VoiceInfo[];
  
  /**
   * Function to load or refresh the available voices
   * @param options Optional request options
   */
  loadVoices: (options?: Partial<VoiceListRequest>) => Promise<VoiceInfo[]>;
}

File: src/backend/schemas/voice.py
from pydantic import BaseModel, Field, validator, root_validator
from typing import Optional, List, Dict, Any, Union
from io import BytesIO
import base64

# Global constants
VOICE_PROVIDERS = ['system', 'elevenlabs', 'coqui']
AUDIO_FORMATS = ['mp3', 'wav', 'ogg', 'flac']
DEFAULT_AUDIO_FORMAT = 'mp3'
DEFAULT_VOICE_PROVIDER = 'elevenlabs'
DEFAULT_STT_MODEL = 'whisper'
MAX_AUDIO_SIZE_MB = 10  # Maximum audio size in MB


class TranscriptionRequest(BaseModel):
    """
    Schema for speech-to-text transcription request.
    
    Either audio_data or audio_file_path must be provided.
    """
    audio_data: Optional[str] = Field(
        None, 
        description="Base64-encoded audio data"
    )
    audio_file_path: Optional[str] = Field(
        None, 
        description="Path to audio file on server"
    )
    language: Optional[str] = Field(
        None, 
        description="Language code (e.g., 'en', 'fr', 'es'); if None, auto-detection is used"
    )
    model: Optional[str] = Field(
        None, 
        description="Speech recognition model to use"
    )
    temperature: Optional[float] = Field(
        None, 
        description="Sampling temperature, between 0 and 1"
    )
    word_timestamps: Optional[bool] = Field(
        False, 
        description="Whether to include word-level timestamps"
    )
    
    @root_validator
    def validate_audio_source(cls, values):
        """Validate that at least one audio source is provided."""
        if values.get('audio_data') is None and values.get('audio_file_path') is None:
            raise ValueError("Either audio_data or audio_file_path must be provided")
        return values
    
    @validator('audio_data')
    def validate_audio_data(cls, audio_data):
        """Validate that audio_data is properly base64 encoded."""
        if audio_data is None:
            return None
        
        try:
            # Try to decode to validate it's proper base64
            base64.b64decode(audio_data)
        except Exception:
            raise ValueError("Invalid base64 encoding for audio_data")
        
        return audio_data
    
    @validator('model')
    def validate_model(cls, model):
        """Validate that the model is supported."""
        if model is None:
            return DEFAULT_STT_MODEL
        
        supported_models = ['whisper', 'whisper-small', 'whisper-medium', 'whisper-large']
        if model not in supported_models:
            raise ValueError(f"Unsupported model. Choose from: {', '.join(supported_models)}")
        
        return model
    
    @validator('temperature')
    def validate_temperature(cls, temperature):
        """Validate that temperature is within acceptable range."""
        if temperature is None:
            return 0.0
        
        if not 0.0 <= temperature <= 1.0:
            raise ValueError("Temperature must be between 0.0 and 1.0")
        
        return temperature


class TranscriptionResponse(BaseModel):
    """
    Schema for speech-to-text transcription response.
    """
    text: str = Field(..., description="Transcribed text")
    confidence: float = Field(..., description="Confidence score between 0 and 1")
    language: str = Field(..., description="Detected or specified language code")
    model: str = Field(..., description="Model used for transcription")
    segments: Optional[Dict[str, Any]] = Field(
        None, 
        description="Word-level segments with timestamps if requested"
    )
    metadata: Optional[Dict[str, Any]] = Field(
        None, 
        description="Additional metadata about the transcription"
    )


class SynthesisRequest(BaseModel):
    """
    Schema for text-to-speech synthesis request.
    """
    text: str = Field(..., description="Text to synthesize")
    voice_id: Optional[str] = Field(
        None, 
        description="Voice ID to use for synthesis"
    )
    provider: Optional[str] = Field(
        None, 
        description="Voice provider to use"
    )
    speed: Optional[float] = Field(
        None, 
        description="Speech speed, between 0.5 and 2.0"
    )
    pitch: Optional[float] = Field(
        None, 
        description="Voice pitch, between 0.5 and 2.0"
    )
    output_format: Optional[str] = Field(
        None, 
        description="Audio output format"
    )
    stream: Optional[bool] = Field(
        None, 
        description="Whether to stream the audio response"
    )
    provider_options: Optional[Dict[str, Any]] = Field(
        None, 
        description="Provider-specific options"
    )
    
    @validator('text')
    def validate_text(cls, text):
        """Validate that text is not empty."""
        if not text or not text.strip():
            raise ValueError("Text cannot be empty")
        return text
    
    @validator('provider')
    def validate_provider(cls, provider):
        """Validate that the provider is supported."""
        if provider is None:
            return DEFAULT_VOICE_PROVIDER
        
        if provider not in VOICE_PROVIDERS:
            raise ValueError(f"Unsupported provider. Choose from: {', '.join(VOICE_PROVIDERS)}")
        
        return provider
    
    @validator('speed')
    def validate_speed(cls, speed):
        """Validate that speed is within acceptable range."""
        if speed is None:
            return 1.0
        
        if not 0.5 <= speed <= 2.0:
            raise ValueError("Speed must be between 0.5 and 2.0")
        
        return speed
    
    @validator('pitch')
    def validate_pitch(cls, pitch):
        """Validate that pitch is within acceptable range."""
        if pitch is None:
            return 1.0
        
        if not 0.5 <= pitch <= 2.0:
            raise ValueError("Pitch must be between 0.5 and 2.0")
        
        return pitch
    
    @validator('output_format')
    def validate_output_format(cls, output_format):
        """Validate that output format is supported."""
        if output_format is None:
            return DEFAULT_AUDIO_FORMAT
        
        if output_format not in AUDIO_FORMATS:
            raise ValueError(f"Unsupported output format. Choose from: {', '.join(AUDIO_FORMATS)}")
        
        return output_format
    
    @validator('stream')
    def validate_stream(cls, stream):
        """Validate stream flag."""
        if stream is None:
            return False
        return stream


class SynthesisResponse(BaseModel):
    """
    Schema for text-to-speech synthesis response.
    """
    audio_data: str = Field(..., description="Base64-encoded audio data")
    content_type: str = Field(..., description="Content type of audio (e.g., 'audio/mp3')")
    provider: str = Field(..., description="Provider used for synthesis")
    voice_id: str = Field(..., description="Voice ID used for synthesis")
    metadata: Optional[Dict[str, Any]] = Field(
        None, 
        description="Additional metadata about the synthesis"
    )
    
    @validator('audio_data')
    def validate_audio_data(cls, audio_data):
        """Validate that audio_data is properly base64 encoded."""
        try:
            # Try to decode to validate it's proper base64
            base64.b64decode(audio_data)
        except Exception:
            raise ValueError("Invalid base64 encoding for audio_data")
        
        return audio_data
    
    @validator('content_type')
    def validate_content_type(cls, content_type):
        """Validate content type format."""
        if not content_type.startswith('audio/'):
            raise ValueError("Content type must be an audio format (e.g., 'audio/mp3')")
        
        return content_type


class VoiceInfo(BaseModel):
    """
    Schema for voice information.
    """
    voice_id: str = Field(..., description="Unique identifier for the voice")
    name: str = Field(..., description="Display name of the voice")
    provider: str = Field(..., description="Provider of the voice")
    gender: Optional[str] = Field(None, description="Gender of the voice (if applicable)")
    language: Optional[str] = Field(None, description="Primary language of the voice")
    accent: Optional[str] = Field(None, description="Accent of the voice (if applicable)")
    description: Optional[str] = Field(None, description="Description of the voice")
    preview_url: Optional[str] = Field(None, description="URL to a preview audio sample")
    additional_info: Optional[Dict[str, Any]] = Field(
        None, 
        description="Additional provider-specific information"
    )
    
    @validator('provider')
    def validate_provider(cls, provider):
        """Validate that the provider is supported."""
        if provider not in VOICE_PROVIDERS:
            raise ValueError(f"Unsupported provider. Choose from: {', '.join(VOICE_PROVIDERS)}")
        
        return provider


class VoiceListRequest(BaseModel):
    """
    Schema for requesting available voices.
    """
    provider: Optional[str] = Field(
        None, 
        description="Provider to filter voices by, or 'all' for all providers"
    )
    force_refresh: Optional[bool] = Field(
        None, 
        description="Whether to force a refresh of the voice list from providers"
    )
    
    @validator('provider')
    def validate_provider(cls, provider):
        """Validate that the provider is supported."""
        if provider is None:
            return 'all'
        
        if provider != 'all' and provider not in VOICE_PROVIDERS:
            raise ValueError(f"Unsupported provider. Choose from: 'all', {', '.join(VOICE_PROVIDERS)}")
        
        return provider
    
    @validator('force_refresh')
    def validate_force_refresh(cls, force_refresh):
        """Validate force_refresh flag."""
        if force_refresh is None:
            return False
        return force_refresh


class VoiceListResponse(BaseModel):
    """
    Schema for response containing available voices.
    """
    voices: List[VoiceInfo] = Field(..., description="List of available voices")
    count_by_provider: Dict[str, int] = Field(
        ..., 
        description="Count of voices per provider"
    )